### PR TITLE
approaching to the functional paradigm (`numberToWords`) + handle the largest possible number (`2^53 - 1`)

### PR DESCRIPTION
**Pull request by parsa-vxt**

### CODE REVIEW

File: src/modules/numberToWords/index.ts [hunk 1]
```diff
@@ -1,132 +1,69 @@
...
-}
+type NumberToWordsType = (inputNumber: number | string, options?: NumberToWordsOptions) => string | TypeError;
 
-const config = {
-	scale: ["", "هزار", "میلیون", "میلیارد"],
-};
+const numberToWords: NumberToWordsType = (inputNumber, options) => {
+	const isNumberValid = (n: number) => typeof n === "number" && Number.isSafeInteger(n) && n !== 0;
+	const isNegative = (n: number) => n < 0;
+	const notValidNumberError = () => TypeError("the number must be a safe integer value");
```
Comment (ali-master) [inline src/modules/numberToWords/index.ts]: TypeError message should follow the PersianTools errors rule while you can see in some other files we've added before.

Comment (ali-master) [reply src/modules/numberToWords/index.ts]: `notValidNumberError` should be `numberIsNotValidError`.

I prefer to create a Class and add all our Errors in one class which helps us to be regular.

File: test/NumberToWords.spec.ts [hunk 2]
```diff
@@ -12,6 +12,8 @@ it("numberToWords", () => {
 	expect(numberToWords(45, { ordinal: true })).toEqual("چهل و پنجم");
 	expect(numberToWords(0)).toEqual("صفر");
 	// @ts-ignore
-	expect(numberToWords()).toEqual("");
-	expect(numberToWords(502375902532527)).toEqual("");
+	expect(numberToWords()).toBeInstanceOf(TypeError);
+	expect(numberToWords(9006199254740992)).toEqual(
```
Comment (ali-master) [inline test/NumberToWords.spec.ts:21]: Add more tests and also update the README to inform users that this function supports large numbers.

File: src/modules/numberToWords/numbersWordList.ts [hunk 3]
```diff
@@ -0,0 +1,49 @@
+const numbersWordList: Map<number, string> = new Map([
```
Comment (ali-master) [inline src/modules/numberToWords/numbersWordList.ts]: Please check the browser compatibility after adding this function to ensure that the bundle works well in the old nodejs and browser versions.

File: src/modules/numberToWords/index.ts [hunk 4]
```diff
@@ -1,132 +1,69 @@
...
-		return "";
+		return numbersArr.join(" و ").trim();
 	}
 
-	for (let i = 0; i < result.length; i++) {
-		if (result[i] !== "") {
-			result[i] += " " + config.scale[i] + " و ";
+	const positiveNumber = Math.abs(number);
+	const handleResult = () => {
+		if (Number(inputNumber) === 0) return "صفر";
```
Comment (ali-master) [inline src/modules/numberToWords/index.ts]: Seems your code is not prettified.

Comment (ali-master) [reply src/modules/numberToWords/index.ts]: The `inputNumber` is a wrong variable name. it should `numberValue` or something like that.

File: src/modules/numberToWords/index.ts [hunk 5]
```diff
@@ -1,132 +1,69 @@
...
- * @category Number conversion
- * @return Converted numbers to words. e.g: صد و بیست و یک
- */
-function numberToWords(number: bigint | number | string, { ordinal = false }: NumberToWordsOptions = {}): string {
-	if (typeof number === "undefined") return "";
 
-	if (number === 0) {
-		return "صفر";
-	}
+	function performer(num: number) {
```
Comment (ali-master) [inline src/modules/numberToWords/index.ts]: What is this function? what does it do? Write document above the function and also add return type.

**Comment by ali-master** [comment]
Thanks for contributing to the Project and I appreciate it.

All your codes and lines should be documented and covered in the tests and please write more and more and more tests. one test is not enough. There are a lot of users who are using our project and we have to be very cautious about every single line of our codes.

**Review by ali-master — CHANGES_REQUESTED** [review]
I reviewed and there are some issues.

**Comment by parsa-vxt** [comment]
thanks for the review, I will fix these issues as soon as possible.

**Comment by parsa-vxt** [comment]
There is a bunch of failed tests that are not related to that part I am working on, so I can't/should not commit...
should I neglect those?

**Comment by ali-master** [comment]
> There is a bunch of failed tests that are not related to that part I am working on, so I can't/should not commit...
> should I neglect those?

I know, you should write at least 10 more tests to make sure that your changes have no backward compatibility side effects in production.

We want to release new version, so please complete your PR as soon as possible.

**Comment by parsa-vxt** [comment]
the issues were resolved completely... BTW sorry for the delay.

**Comment by ali-master** [comment]
@all-contributors please add @psparsa for tests, code and ideas

**Comment by ali-master** [comment]
@psparsa Welcome to the PersianTools family. We are glad that you are a member of our family.

**Pull request merged**